index.js

```diff
@@ -1,98 +1,187 @@
+  var transport = this._transport;
+
+  var promise;
+  if(this._isOpen){
+    promise = when.reject(new Error('Transport already open.'));
+  } else {
+    // the close method removes all event listeners,
+    // so we need to rebind on our open
+    transport.on('data', function(chunk){
+      if(typeof self._queue === 'function'){
+        self._queue(chunk);
+      }
+    });
 
-  function _open(){
-    return when.promise(function(resolve, reject) {
-      serialport.open( function(err){
-        if(err){ return reject(err); }
-        return resolve();
+    promise = transport.open()
+      .tap(function(){
+        self._isOpen = true;
       });
-    });
   }
 
-  function onChunk(data) {
-    self.emit('data', data);
+  return nodefn.bindCallback(promise, cb);
+};
+
+Protocol.prototype._close = function(cb){
+  var self = this;
+  var transport = this._transport;
+
+  function onClose(){
+    self._isOpen = false;
   }
 
-  self._onResponse(onChunk);
+  var promise = transport.close()
+    .tap(onClose)
+    .catch(function(err){
+      if(err && !openRegexp.test(err.message)){
+        // rethrow error if it is not "Serialport not open."
+        throw err;
+      }
+
+      onClose();
+    });
 
-  var promise = _open();
   return nodefn.bindCallback(promise, cb);
 };
 
-Protocol.prototype.close = function(cb){
-  var serialport = this._serial;
+Protocol.prototype._setDtr = function(cb){
+  var transport = this._transport;
 
-  function _close(){
-    return when.promise(function(resolve, reject) {
-      serialport.close( function(err){
-        if(err){ return reject(err); }
-        return resolve();
-      });
+  var promise = transport.set({ dtr: false });
+
+  return nodefn.bindCallback(promise, cb);
+};
+
+Protocol.prototype._clrDtr = function(cb){
+  var transport = this._transport;
+
+  var promise = transport.set({ dtr: true });
+
+  return nodefn.bindCallback(promise, cb);
+};
+
+Protocol.prototype._setBrk = function(cb){
+  var transport = this._transport;
+
+  var brkBit = new Buffer([0x00]);
+
+  var promise = transport.update({ baudRate: 200 })
+    .then(function(){
+      return transport.write(brkBit);
     });
-  }
 
-  var promise = this.signoff()
-    .then(_close());
+  return nodefn.bindCallback(promise, cb);
+};
+
+Protocol.prototype._clrBrk = function(cb){
+  var transport = this._transport;
+  var options = this._options.options;
+
+  var promise = transport.update({ baudRate: options.baudrate });
+
   return nodefn.bindCallback(promise, cb);
 };
 
 Protocol.prototype.enterProgramming = function(cb){
-  var serialport = this._serial;
+  var self = this;
 
-  function _open(){
-    return when.promise(function(resolve, reject) {
-      serialport.open( function(err){
-        if(err){ return reject(err); }
-        return resolve();
-      });
+  var promise = this._originalTransportClosed
+    .then(function(){
+      return self._open();
+    })
+    .then(function(){
+      return self._setBrk();
```

Comment:
is it correct to move `setBrk` and `clrBrk` into `enterProgramming` instead of being in `reset`?